src/front/spv/mod.rs

```diff
@@ -1984,6 +1990,90 @@ impl<I: Iterator<Item = u32>> Parser<I> {
                 Op::Fwidth | Op::FwidthFine | Op::FwidthCoarse => {
                     self.parse_expr_derivative(expressions, crate::DerivativeAxis::Width)?;
                 }
+                Op::ArrayLength => {
+                    inst.expect(5)?;
+                    let result_type_id = self.next()?;
+                    let result_id = self.next()?;
+                    let structure_id = self.next()?;
+                    let member_index = self.next()?;
+
+                    // Validate that all the types are correct.
+                    let result_type = self.lookup_type.lookup(result_type_id)?;
+                    match type_arena[result_type.handle].inner {
+                        crate::TypeInner::Scalar {
+                            kind: crate::ScalarKind::Uint,
+                            width: 4,
+                        } => {}
+                        _ => return Err(Error::InvalidParameter(Op::ArrayLength)),
+                    }
+                    let structure = self.lookup_expression.lookup(structure_id)?;
+                    let structure_type = self.lookup_type.lookup(structure.type_id)?;
+                    match type_arena[structure_type.handle].inner {
+                        crate::TypeInner::Pointer { base, .. } => {
+                            if let crate::TypeInner::Struct { ref members, .. } = type_arena[base].inner {
+                                if let Some(last) = members.last() {
+                                    match type_arena[last.ty].inner {
+                                        crate::TypeInner::Array { size: crate::ArraySize::Dynamic, .. } => if members.len() - 1 == member_index as _ {},
+                                        _ => return Err(Error::InvalidParameter(Op::ArrayLength)),
+                                    }
+                                } else {
+                                    return Err(Error::InvalidParameter(Op::ArrayLength))
+                                }
+                            } else {
+                                return Err(Error::InvalidParameter(Op::ArrayLength))
+                            }
+                        }
+                        _ => return Err(Error::InvalidParameter(Op::ArrayLength)),
+                    }
+
+                    let structure_expr = expressions.append(crate::Expression::Load {
```

Comment:
This is incorrect. We shouldn't load anything for determining the array length. It should be:
`ArrayLength -> Access -> GlobalVariable`

Reply:
Does using `Expression::Access` on a pointer to a structure get a pointer to the field?

Reply:
Since `OpArrayLength` in spirv is given a pointer to the structure, maybe it makes sense to stick with that convention throughout naga?

Reply:
> Does using Expression::Access on a pointer to a structure get a pointer to the field?

Yes.

> Since OpArrayLength in spirv is given a pointer to the structure, maybe it makes sense to stick with that convention throughout naga?

Thinking about this... I'm not seeing any issues with going this way. The concerns were 1) implementations complexity, 2) ability to have it on fixed-size arrays.

So for (1) we haven't implemented it on HLSL/MSL, but it seems like that code is not going to be any different. WGSL in/out would be a little bit more complex, SPIR-V in/out would be less complex. So all in all it's the same.

For (2), the frontend can just see the fixed-size array, if it needs to handle it, and spew `Expression::Constant` instead of `Expression::ArrayLength`.

So TL:DR; it seems fine to proceed with the SPIRV semantics here.

Reply:
I decided that I'd prefer if `Expression::ArrayLength` is passed a pointer to the array itself.